KAFKA-9754: Add an option to ignore ProduceBench errors

scott-hendricks opened the pull request:
This change adds the following features to the Trogdor ProduceBenchWorker:

Ability to ignore errors caused by SendRecords.
Tracking of the number of errors ignored.
Additional tracking of the number of callbacks with exceptions.
Ability to extend ProduceBenchWorker's verify phase to deal with large topics.

In order to maintain backward compatibility, ignoring errors is disabled by default.

All build-time integration tests passed:

BUILD SUCCESSFUL in 42m 8s
334 actionable tasks: 255 executed, 79 up-to-date

scott-hendricks committed:
KAFKA-9754 - Add option to ignore producer errors and better error statistics.

scott-hendricks committed:
Add additional configuration options to avoid errors.

scott-hendricks commented:
I have run several tests:

1. Testing with a high-partition count and multiple producers.  I ran with topicVerificationRetries=100 and ignoreProduceErrors=true.  This is the situation that is causing us pain, and the test had no errors.
2. I ran a test that would cause callback errors and verified the counts are increasing properly.
3. I ran a test where the new parameters were not provided, and the trogdor workload behaved identical as before.

andrewegel commented on tools/src/test/java/org/apache/kafka/trogdor/common/WorkerUtilsTest.java:
```diff
@@ -73,7 +73,7 @@ public void setUp() throws Exception {
         Map<String, NewTopic> newTopics = Collections.singletonMap(TEST_TOPIC, NEW_TEST_TOPIC);
 
-        WorkerUtils.createTopics(log, adminClient, newTopics, true);
+        WorkerUtils.createTopics(log, adminClient, newTopics, true, 3, 2500);
```
NIT: Reuse `DEFAULT_TOPIC_VERIFY_BACKOFF` in all `2500` places that you defined?

scott-hendricks replied:
That's a good point.  I'll do that.

scott-hendricks replied:
I also added a single source of truth for the number of retries too.

andrewegel commented on tools/src/test/java/org/apache/kafka/trogdor/common/WorkerUtilsTest.java:
```diff
@@ -90,7 +90,7 @@ public void testCreateRetriesOnTimeout() throws Throwable {
 
         WorkerUtils.createTopics(
-            log, adminClient, Collections.singletonMap(TEST_TOPIC, NEW_TEST_TOPIC), true);
+            log, adminClient, Collections.singletonMap(TEST_TOPIC, NEW_TEST_TOPIC), true, 3, 2500);
```
same about reusing `DEFAULT_TOPIC_VERIFY_BACKOFF`?

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/common/WorkerUtils.java:
```diff
@@ -128,12 +129,38 @@ public static void createTopics(
         Logger log, String bootstrapServers, Map<String, String> commonClientConf,
         Map<String, String> adminClientConf,
         Map<String, NewTopic> topics, boolean failOnExisting) throws Throwable {
+        // This method allows backward compatibility with the original signature of createTopics.
```
We don't need backwards compatibility with the original signature of createTopics.  It's not a public API

scott-hendricks replied:
Ok, I'll remove it then.  I was nervous that this change could potentially break forks unknown to me.

scott-hendricks committed:
Single source of truth, Remove old function signature.

andrewegel reviewed (approved):
LGTM then - Please get an actual Kafka maintainers sign off.

cmccabe commented:
@scott-hendricks , thanks for working on this.  What is the error that you're trying to ignore here?  If it's an error where the produce buffer(s) are full, shouldn't we just wait and retry?  That would be a better solution since it wouldn't involve configuration, and possibly ignoring other errors that don't have a resolution that requires retrying.

scott-hendricks commented:
@cmccabe - In a combination of both the way Trogdor is coded and how we schedule it, we have a race condition when starting up.  The first ProduceBenchWorker to start creates the topic, and the rest of them just verify the topic before proceeding.  If the topic is sufficiently large enough, the secondary ProduceBenchWorker jobs time out waiting on topic verification, failing with `This server does not host this topic-partition.`

I introduced task spacing in our schedule generator a while back to try and combat this, and it works well for small topics.  However it is flaky and does not work well with large topics which, depending on conditions, can take wildly varying times to create.  This change allows us to modify the number of retries in the topic verifier, and we can effectively wait an arbitrary amount of time for the topic to create.

In addition, we have seen issues where we aren't properly counting all the failed SendRecords callback commands, so I have added counters there.  And, while we haven't seen any issues with it yet, when we start to ramp up our fault injection testing I want to be prepared for any errors in the initial SendRecords command, so I added error ignore and counting code there.

cmccabe commented:
If the goal is to wait longer for the topic to be created, shouldn't we make that configurable rather than ignoring all produce errors?  We could just make the time to wait for topic creation configurable and crank it up to a very long value when testing.

scott-hendricks commented:
I ended up implementing both.  If you notice, the PR now contains code to increase the wait time for topic verify as well.

cmccabe commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchSpec.java:
```diff
@@ -72,6 +75,8 @@
     private final TopicsSpec inactiveTopics;
     private final boolean useConfiguredPartitioner;
+    private final boolean ignoreProduceErrors;
+    private final int topicVerificationRetries;
```
It would be better to call this "topicVerificationTries" since 0 is then an unreasonable value.  For REtries, 0 is a reasonable value and not something that we should assume is equivalent to "use the default."
